### DISCUSSION TITLE
How are integrity checks work when uploading objects?

### DISCUSSION THREAD

**Discussion by FKupsch1982** [seaweedfs/seaweedfs#3908] [Q&A] [answered]
Hi, I try to use the [PutObject Rest API](https://docs.aws.amazon.com/AmazonS3/latest/API/API_PutObject.html) with the Content-MD5 or the x-amz-checksum-xy headers to check the integrity of the uploaded data, but I dont see that they are recognized. Are they even supported? 

Regards,
Felix

**Comment by kmlebedev** [comment, marked as answer]
>Additionally, you can calculate the MD5 while putting an object to Amazon S3 and compare the returned ETag to the calculated MD5 value.

**Comment by FKupsch1982** [comment]
Thank you for this hint. Nevertheless, I would be interested to know what exactly is done with the Content-MD5 header. I get an error message if the value of the header is not a valid MD5 checksum, e.g. because it is too short. But if I send an MD5 checksum that does not match the value, it is ignored. Dont think that this is conformant S3 behavior.

**Reply by kmlebedev** [reply]
This implementation of s3 can split the file into chunks on the filer side and store it in blob storage in parallel, and it is not possible to calculate the md5 of the whole file.  But you can calculate the md5 size of the data transferred to the filler, but there is no special point in this and only slow down the work of s3.

**Reply by terribleplan** [reply]
When doing a simple `PutObject` the filer receives the entire file contents in one HTTP request body, and as such should be able to calculate checksums as the upload streams in regardless of how it handles writing the file. In a proper implementation of this if pieces are already written when the file is fully "uploaded"/buffered/whatever they should be deleted and the error reported.

This would not be true in the case of an s3 multipart upload.

**Comment by terribleplan** [comment]
TL;DR: Those headers are not supported.

`X-Amz-Checksum-*` headers are not handled at all.

The `Content-MD5` header [is not used to validate the body of a PutObject operation](https://github.com/seaweedfs/seaweedfs/blob/2e78a522ab3892debf9aefeb978177678ae52a83/weed/s3api/s3api_object_handlers.go#L51), and even the validation of the header [only checks if the header's value is not an empty string and is base64](https://github.com/seaweedfs/seaweedfs/blob/2e78a522ab3892debf9aefeb978177678ae52a83/weed/s3api/s3api_handlers.go#L43), not that it is a 16-byte md5 hash.

**Comment by michalszynkiewicz** [comment]
I was testing s3 api compatibility for a few solutions, and my tests found this feature is not implemented in seaweedfs, so I let myself create a PR to fix it: https://github.com/seaweedfs/seaweedfs/pull/8367

MD5's were already computed for the contents, so I just compared them with the checksum that comes in with the request (if specified)

**Reply by chrislusf** [reply]
@michalszynkiewicz Merged. Thanks!!!
